kotlinx-coroutines-core/common/src/internal/CoroutineExceptionHandlerImpl.kt

```diff
@@ -0,0 +1,94 @@
+ * Once a new callback is registered, it tries.
+ *
+ * The callbacks in this object are the last resort before relying on platform-dependent
+ * ways to report uncaught exceptions from coroutines.
+ */
+@PublishedApi
+internal object ExceptionCollector {
+    private val lock = SynchronizedObject()
+    private var enabled = false
+    private var unprocessedExceptions = mutableListOf<Throwable>()
+    private val callbacks = mutableMapOf<Any, (Throwable) -> Unit>()
+
+    /**
+     * Registers [callback] to be executed when an uncaught exception happens.
+     * [owner] is a key by which to distinguish different callbacks.
+     */
+    fun addOnExceptionCallback(owner: Any, callback: (Throwable) -> Unit) = synchronized(lock) {
+        enabled = true // never becomes `false` again
+        val previousValue = callbacks.put(owner, callback)
+        assert { previousValue === null }
+        // try to process the exceptions using the newly-registered callback
+        unprocessedExceptions.forEach { reportException(it) }
+        unprocessedExceptions = mutableListOf()
+    }
+
+    /**
+     * Unregisters the callback associated with [owner].
+     */
+    fun removeOnExceptionCallback(owner: Any) = synchronized(lock) {
+        val existingValue = callbacks.remove(owner)
+        assert { existingValue !== null }
+    }
+
+    /**
+     * Tries to handle the exception by propagating it to an interested consumer.
+     * Returns `true` if the exception does not need further processing.
+     *
+     * Doesn't throw.
+     */
+    fun handleException(exception: Throwable): Boolean = synchronized(lock) {
+        if (!enabled) return false
+        if (reportException(exception)) return true
+        /** we don't return the result of the `add` function because we don't have a guarantee
+         * that a callback will eventually appear and collect the unprocessed exceptions, so
+         * we can't consider [exception] to be properly handled. */
+        unprocessedExceptions.add(exception)
+        return false
+    }
+
+    /**
+     * Try to report [exception] to the existing callbacks.
+     */
+    private fun reportException(exception: Throwable): Boolean {
+        var executedACallback = false
+        for (callback in callbacks.values) {
+            callback(exception)
+            executedACallback = true
+            /** We don't leave the function here because we want to fan-out the exceptions to every interested consumer,
+             * it's not enough to have the exception processed by one of them.
+             * The reason is, it's less big of a deal to observe multiple concurrent reports of bad behavior than not
+             * to observe the report in the exact callback that is connected to that bad behavior. */
+        }
+        return executedACallback
+    }
+}
+
+internal fun handleUncaughtCoroutineException(context: CoroutineContext, exception: Throwable) {
+    if (ANDROID_DETECTED) {
```

Comment:
I got the idea behind this guard-check, but it still confuses me.

Can we instead guard it by, forgive me, Father, global mutable atomic boolean that the test module switches on? 
In addition, we can add `assumenosideeffects EXCEPTIONS_TEST_MODULE_FLAG false` to save Android a few bytes in dex

Reply:
Do the proguard rules work correctly with atomic values?

Also, the guard check is needed solely for the Android to save a few bytes, it has no significance other than that. I dropped a comment to clarify this.

Reply:
They don't, but you do not have to expose the atomic, it can be a property with custom get/set and atomic as backing field

Reply:
Could you please add a corresponding check to our r8 test?
It probably should be `-checkdiscard kotlinx.coroutines.internal.ExceptionCollector`